features/lorawan/LoRaWANStack.cpp

```diff
@@ -0,0 +1,2066 @@
+#define LC6                { 867500000, 0, { ( ( DR_5 << 4 ) | DR_0 ) }, 0 }
+#define LC7                { 867700000, 0, { ( ( DR_5 << 4 ) | DR_0 ) }, 0 }
+#define LC8                { 867900000, 0, { ( ( DR_5 << 4 ) | DR_0 ) }, 0 }
+#define LC9                { 868800000, 0, { ( ( DR_7 << 4 ) | DR_7 ) }, 2 }
+#define LC10               { 868300000, 0, { ( ( DR_6 << 4 ) | DR_6 ) }, 1 }
+#endif
+#endif
+
+/**
+ *
+ * User application data buffer size if compliance test is used
+ */
+#if (MBED_CONF_LORA_PHY == 0 || MBED_CONF_LORA_PHY == 4 || MBED_CONF_LORA_PHY == 6 || MBED_CONF_LORA_PHY == 7)
+#define LORAWAN_COMPLIANCE_TEST_DATA_SIZE                  16
+#elif (MBED_CONF_LORA_PHY == 1 || MBED_CONF_LORA_PHY == 2 || MBED_CONF_LORA_PHY == 8 || MBED_CONF_LORA_PHY == 9)
+#define LORAWAN_COMPLIANCE_TEST_DATA_SIZE                  11
+#else
+#error "Must set LoRa PHY layer parameters."
+#endif
+
+/*****************************************************************************
+ * Private Member Functions                                                  *
+ ****************************************************************************/
+bool LoRaWANStack::is_port_valid(uint8_t port)
+{
+    //Application should not use reserved and illegal port numbers.
+    if (port >= 224 || port == 0) {
+        return false;
+    } else {
+        return true;
+    }
+}
+
+lora_mac_status_t LoRaWANStack::set_application_port(uint8_t port)
+{
+    if (is_port_valid(port)) {
+        _app_port = port;
+        return LORA_MAC_STATUS_OK;
+    }
+
+    return LORA_MAC_STATUS_PORT_INVALID;
+}
+
+/*****************************************************************************
+ * Constructor and destructor                                                *
+ ****************************************************************************/
+LoRaWANStack::LoRaWANStack()
+: _device_current_state(DEVICE_STATE_NOT_INITIALIZED), _mac_handlers(NULL),
+  _num_retry(1), _duty_cycle_on(LORAWAN_DUTYCYCLE_ON)
+{
+#ifdef MBED_CONF_LORA_APP_PORT
+    // is_port_valid() is not virtual, so we can call it in constructor
+    if (is_port_valid(MBED_CONF_LORA_APP_PORT)) {
+        _app_port = MBED_CONF_LORA_APP_PORT;
+    } else {
+        tr_error("User defined port in .json is illegal.");
+        _app_port = INVALID_PORT;
+    }
+
+#else
+    // initialize it to INVALID_PORT (255) an illegal port number.
+    // user should set the port
+     _app_port = INVALID_PORT;
+#endif
+
+     memset(&_lw_session, 0, sizeof(_lw_session));
+     memset(&_tx_msg, 0, sizeof(_tx_msg));
+     memset(&_rx_msg, 0, sizeof(_rx_msg));
+}
+
+LoRaWANStack::~LoRaWANStack()
+{
+
+}
+
+/*****************************************************************************
+ * Public member functions                                                   *
+ ****************************************************************************/
+LoRaWANStack& LoRaWANStack::get_lorawan_stack()
+{
+    static LoRaWANStack _lw_stack;
```

Comment:
I am not certain if this is thead safe in the C++ version we use (C++1x should be). Should it be? I would assume yes, at least the construction of  this stack to be thread safe.

Reply:
At the moment, LoRaWANStack which is the controller layer for the Mac and Phy works in the main thread, as explained in a comment above, we are moving towards running everything in main thread  except the radio driver. In addition to that Mac layer, which is a monolith at the moment, will be torn apart and re-implemented as a bunch of helper classes. All of this was mentioned in the document I shared last week. So this class being a Singleton, will not be valid anymore.